### PR TITLE
Move keymap to separate repository

### LINKED ISSUE #636: Move keymap to a separate repository
I'd like to move libraries/keymap to a seperate repository and publish it on Quicklisp.
@jmercouris Where shall we host it?
I can host it myself, or have it here at github.com/atlas-engineer.

### PR DESCRIPTION
**Pull request by aadcg**
# Description

Fixes #636

# Discussion

Needs to be tested! Can only be merged once the patch is accepted by Guix.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.  

- [ ] I have pulled from master before submitting this PR
- [ ] My code follows the style guidelines for Common Lisp code
  - See [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf) and [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml).
- [ ] I have performed a self-review of my own code
- [ ] My code has been reviewed by at least one peer (the peer review to approve a PR counts.  The reviewer must download and test the code)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] There are no merge conflicts

### CODE REVIEW

**Comment by Ambrevar** [comment]
1. How did you add the git submodule?  If manually, it won't work and it seems that the submodule commit is missing indeed.  To properly add a git submodule, call `git submodule ... REPO-URL`, then adjust the `.gitmodule` file and finally push this file _together with_ the submodule commit.
2. Should we switch to the `keymaps` package?  Maybe, otherwise we could also keep compatibility with the previous one, but considering we are going to break the API anyways... In all cases, you must update the package local nicknames, else it won't work.  See `source/package.lisp`.

**Comment by aadcg** [comment]
Regarding 2, I forgot to push that commit!  My rationale was to keep the parallel with `nhooks` and `nfiles`. Perhaps we should add the nickname `keymap` and deprecate it in version 4?

Will address 1.

**Comment by aadcg** [comment]
This should be good now. 

With respect to keymaps, I think the only thing left is to do the renaming as discussed in #1915. Should I do it @Ambrevar?

**Comment by Ambrevar** [comment]
Nope, see https://github.com/atlas-engineer/nkeymaps/pull/2.

**Comment by Ambrevar** [comment]
I've upated the whole repo to follow the nkeymaps 1-series API.

I _have not_ updated the Guix package nor the submodule yet, so you'll need a local checkout to test this.

We can update them when we agree to merge 1-series on nkeymaps' `master`.

@jmercouris @aadcg @aartaka What do you think of the new API?

**Comment by aadcg** [comment]
It seems to be working fine. I've pushed minor corrections.

I think there are still some concerns to address:

- Rename the slots from `keymap-scheme` to `keybindings`
- Is `keymap-scheme-name` a good name (see `input-buffer`)?
- Shouldn't the slot `scheme-name` from `keymap-scheme-mode` be renamed to `keyscheme`?
- Re-read `scheme.lisp` and `mode/keymap-scheme.lisp` to make sure it all makes sense.

**Comment by aadcg** [comment]
@Ambrevar, please review the last 3 commits I've pushed.

File: source/mode/keymap-scheme.lisp [hunk 1]
```diff
@@ -11,10 +11,12 @@ Ensures that a single keybindings mode, such as `nyxt/emacs-mode', is enabled.")
   "All modes that set `keymap-scheme-name' should inherit from this mode.
 Ensures that a single keybindings mode, such as `nyxt/emacs-mode', is enabled."
   ((rememberable-p nil)
+   ;; TODO Rename to keyscheme?
    (scheme-name
     keyscheme:cua
     :type keymaps:keyscheme
     :documentation "The `keymaps:keyscheme' to enable.")
+   ;; TODO Rename to previous-keyscheme?
```
Comment (Ambrevar) [inline source/mode/keymap-scheme.lisp:19]: Yes to both!

File: source/mode/os-package-manager.lisp [hunk 2]
```diff
@@ -16,14 +16,14 @@
    (current-process-info nil
                          :type (or null uiop/launch-program::process-info))
    (keymap-scheme
```
Comment (Ambrevar) [inline source/mode/os-package-manager.lisp:18]: Should be `keyscheme-map`.

**Comment by Ambrevar** [comment]
> * Rename the slots from `keymap-scheme` to `keybindings`

Why not `keyscheme-map` to be consistent?

>     * Is `keymap-scheme-name` a good name (see `input-buffer`)?

No, scheme-name no longer exists, so it should be `keyscheme` here.

>     * Shouldn't the slot `scheme-name` from `keymap-scheme-mode` be renamed to `keyscheme`?

Yes!

>     * Re-read `scheme.lisp` and `mode/keymap-scheme.lisp` to make sure it all makes sense.

I've left some comments.

**Comment by Ambrevar** [comment]
In the end I've updated the nkeymaps submodule to track the 1-series branch so that we can get feedback from the CI.

Note that I was not able to update the branch of the submodule, I had to delete the module and add it again with

```sh
git submodule add -b 1-series https://github.com/atlas-engineer/nkeymaps _build/nkeymaps
```

File: source/mode/keymap-scheme.lisp [hunk 3]
```diff
@@ -2,22 +2,25 @@
 ;;;; SPDX-License-Identifier: BSD-3-Clause
 
 (nyxt:define-package :nyxt/keymap-scheme-mode
```
[the remark this answers is not in the document]
Comment (aadcg) [reply source/mode/keymap-scheme.lisp:4]: Absolutely!

**Comment by aadcg** [comment]
@Ambrevar I agree with all of your input and made the changes accordingly. Please review it, as I have found myself mixing the concepts at times!

There's a problem though. Just try to compile it and you'll see it. I don't get it, since all we've done was renaming...

**Comment by Ambrevar** [comment]
Fixed. See my last commit for an explanation.

**Comment by Ambrevar** [comment]
Guix package updated.
I'll merge this tomorrow.
@aadcg @aartaka Once merged, either `guix pull` or clone nkeymaps to your `~/common-lisp`.

**Comment by aadcg** [comment]
All clear. Thanks @Ambrevar.

**Comment by Ambrevar** [comment]
Merged with afe1bda4246c11f61aa42f7bca179c89d691e2b0.

Thanks a lot for your help, @aadcg !

**Pull request closed without merging**